src/mlpack/methods/kde/kde_stat.hpp

```diff
@@ -48,6 +49,128 @@ class KDEStat
+        }
+        else
+        {
+          const arma::mat& U = statChild0.eigVec;
+          const arma::mat& V = statChild1.eigVec;
+          const arma::vec& mean0 = statChild0.mean;
+          const arma::vec& mean1 = statChild1.mean;
+          const arma::vec& eigVal0 = statChild0.eigVal;
+          const arma::vec& eigVal1 = statChild1.eigVal;
+
+          // Orthonormal basis.
+          const arma::mat G = U.t() * V;
+          arma::mat H = V - U * G;
+          //arma::mat h = mean0 - U * U.t() * (mean0 - mean1);
+          H.insert_cols(H.n_cols, mean0 - U * U.t() * (mean0 - mean1));
+          const arma::mat v = arma::orth(H);
+
+          // New eigenproblem.
+          const size_t N = child0.NumDescendants();
+          const size_t M = child1.NumDescendants();
+          const size_t P = N + M;
+          const size_t s = eigVal0.size() + v.n_cols;
+          const arma::mat gamma = v.t() * V;
+          arma::mat firstMat(s, s, arma::fill::zeros);
+          firstMat.submat(0, 0, eigVal0.size() - 1, eigVal0.size() - 1) =
+              arma::diagmat(eigVal0);
+
+          arma::mat secondMat(s, s, arma::fill::zeros);
+          // Upper left.
+          secondMat.submat(0, 0, G.n_rows - 1, G.n_rows - 1) =
+              G * arma::diagmat(eigVal1) * G.t();
+          // Lower left.
+          secondMat.submat(G.n_rows, 0, s - 1, G.n_rows - 1) =
+              gamma * arma::diagmat(eigVal1) * G.t();
+          // Upper right.
+          secondMat.submat(0, G.n_rows, G.n_rows - 1, s - 1) =
+              G * arma::diagmat(eigVal1) * gamma.t();
+          // Lower right.
+          secondMat.submat(G.n_rows, G.n_rows, s - 1, s - 1) =
+              gamma * arma::diagmat(eigVal1) * gamma.t();
+
+          arma::mat thirdMat(s, s, arma::fill::zeros);
+          // Upper left.
+          thirdMat.submat(0, 0, G.n_rows - 1, G.n_rows - 1) = G * G.t();
+          // Lower left.
+          thirdMat.submat(G.n_rows, 0, s - 1, G.n_rows - 1) = gamma * G.t();
+          // Upper right.
+          thirdMat.submat(0, G.n_rows, G.n_rows - 1, s - 1) = G * gamma.t();
+          // Lower right.
+          thirdMat.submat(G.n_rows, G.n_rows, s - 1, s - 1) = gamma * gamma.t();
+
+          arma::mat newEigenProblem = ((double) N / P) * firstMat +
+                                      ((double) M / P) * secondMat +
+                                      ((double) N * M / (P * P)) * thirdMat;
+
+          // Right singular values. Unused.
+          arma::mat rightEigVal, newEigVec;
+          arma::vec newEigVal;
+
+          arma::svd(newEigVec, newEigVal, rightEigVal, newEigenProblem);
+
+          arma::mat upsilon(U.n_rows, U.n_cols + v.n_cols);
+          upsilon.cols(0, U.n_cols - 1) = U;
+          upsilon.cols(U.n_cols, upsilon.n_cols - 1) = v;
+
+          // Update node's orthogonal base.
+          mean = (1 / (double) P) * (N * mean0 + M * mean1);
+          eigVal = newEigVal;
+          eigVec = upsilon * newEigVec;
```

Comment:
I haven't read the eigenspaces merging paper yet so I can't comment on this, but I'll review it more closely later.